### PR TITLE
Nav bar

### LINKED ISSUE #346: Navigation Bar
Add a navigation bar for the site.

### PR DESCRIPTION
**Pull request by jdoxley**
Make sure you familiarize yourself with our contributing guidelines.

# Describe the PR
Add a main nav bar to the site

# Closes
Closes #346 

# Changes Made
Too many to list.

# Notes
Doesn't have any links by default yet.

### CODE REVIEW

**Comment by jdoxley** [comment]
@Soyvolon Conflicts in program and imports can be resolved by just combining the two. I dont know about the db context thats all you.

File: ProjectDataCore.Components/Nav/EditModuleDisplay.razor [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        _navModuleService.OnDblClickTrigger(this, m);
+    }
+
+    void OnLeftClick(NavModule m)
+    {
+        _navModuleService.OnLeftClickTrigger(this, m);
+    }
+
+}
```
Comment (Soyvolon) [inline ProjectDataCore.Components/Nav/EditModuleDisplay.razor:28]: All code should be in code-behind files. You can use quick actions and refactoring to move the `@code` block into one of those files automatically.

File: ProjectDataCore.Data/Services/Nav/INavModuleService.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    public event EventHandler<NavModule> OnDblClick;
+    public event EventHandler<NavModule> OnLeftClick;
+    public void OnLeftClickTrigger(object? sender, NavModule module);
+    public void OnDblClickTrigger(object? sender, NavModule module);
+    public Task<ActionResult> CreateNavModuleAsync(string displayName, string href, bool hasMainPage, Guid? parent = null);
+    public Task<ActionResult> CreateNavModuleAsync(NavModule module);
+    public Task<List<NavModule>> GetAllModules();
+    public Task<List<NavModule>> GetAllModulesWithChildren();
+    Task<ActionResult> UpdateNavModuleAsync(NavModule navModule);
+    Task<ActionResult> DeleteNavModule(Guid key);
```
Comment (Soyvolon) [inline ProjectDataCore.Data/Services/Nav/INavModuleService.cs:16]: Make sure these are marked `public`

File: ProjectDataCore.Data/Services/Nav/NavModuleService.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
+﻿using ProjectDataCore.Data.Structures.Nav;
+
+namespace ProjectDataCore.Data.Services.Nav;
+
+public class NavModuleService : INavModuleService
+{
+    private readonly IDbContextFactory<ApplicationDbContext> _dbContextFactory;
+
+    public event EventHandler<NavModule> OnDblClick;
+    public event EventHandler<NavModule> OnLeftClick;
```
Comment (Soyvolon) [inline ProjectDataCore.Data/Services/Nav/NavModuleService.cs:10]: Why are these here and not in the component that utilizes them?

File: ProjectDataCore.Data/Structures/Nav/NavModule.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        Href = href;
+        SubModules = subModules;
+        HasMainPage = hasMainPage;
+    }
+    public string DisplayName { get; set; }
+    public string? Href { get; set; }
+    public List<NavModule>? SubModules { get; set; }
+    public bool HasMainPage { get; set; } = false;
+    public NavModule? Parent { get; set; }
+    public Guid? ParentId { get; set; }
```
Comment (Soyvolon) [inline ProjectDataCore.Data/Structures/Nav/NavModule.cs:40]: Please place these above the constructors.

File: ProjectDataCore/Pages/Admin/NavBar/ModifyNavBar.razor [hunk 5]
```diff
@@ -0,0 +1,51 @@
+﻿@page "/admin/navbar/edit"
+@using ProjectDataCore.Data.Structures.Nav
```
Comment (Soyvolon) [inline ProjectDataCore/Pages/Admin/NavBar/ModifyNavBar.razor:2]: If this is used in more than one place put it into the imports file, please.

File: ProjectDataCore/Pages/Admin/NavBar/ModifyNavBar.razor.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
+﻿using ProjectDataCore.Data.Services.Nav;
+using ProjectDataCore.Data.Structures.Nav;
+
+namespace ProjectDataCore.Pages.Admin.NavBar
+{
+    public partial class ModifyNavBar : ComponentBase, IDisposable
+    {
+        [Inject] private INavModuleService _navModuleService { get; set; }
```
Comment (Soyvolon) [inline ProjectDataCore/Pages/Admin/NavBar/ModifyNavBar.razor.cs:8]: For consistency, make this a public property, please (and the name formatting that goes with it).

File: ProjectDataCore/Pages/Admin/NavBar/ModifyNavBar.razor.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            await _navModuleService.DeleteNavModule(editing.Key);
+            editing = null;
+            navModules = await _navModuleService.GetAllModulesWithChildren();
+            await InvokeAsync(StateHasChanged);
+        }
+
+        public void Dispose()
+        {
+            _navModuleService.OnDblClick -= EditNavModule;
+            _navModuleService.OnLeftClick += NewNavModule;
```
Comment (Soyvolon) [inline ProjectDataCore/Pages/Admin/NavBar/ModifyNavBar.razor.cs:84]: You should **never** assign to events in the dispose statement. This only occurs when you are unloading this component *completely*, so there is absolutely no reason to register a new event which will then lead to a disposed object. This will lead to a memory leak as events build up.

File: ProjectDataCore/Shared/ModuleDisplay.razor [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    public string URI { get; set; }
+
+    [Inject]
+    public NavigationManager _navManager { get; set; }
+
+    private void Navigate(string href)
+    {
+        _navManager.NavigateTo(href, true);
+    }
+}
```
Comment (Soyvolon) [inline ProjectDataCore/Shared/ModuleDisplay.razor:75]: Code-behind file.

File: ProjectDataCore/Shared/NavMenu.razor [hunk 9]
```diff
@@ -0,0 +1,63 @@
+﻿@implements IDisposable
```
Comment (Soyvolon) [inline ProjectDataCore/Shared/NavMenu.razor:1]: This can exist only in the code behind and work fine.

File: ProjectDataCore/Shared/NavMenu.razor.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+namespace ProjectDataCore.Shared;
+
+public partial class NavMenu : ComponentBase, IDisposable
+{
+    private List<NavModule> _modules = new List<NavModule>();
+
+    public string URI { get; set; } = "";
+
+    [Inject] public NavigationManager _navManager { get; set; }
+    [Inject] public INavModuleService _navModuleService { get; set; }
```
Comment (Soyvolon) [inline ProjectDataCore/Shared/NavMenu.razor.cs:14]: Make these public please.

**Review by Soyvolon — CHANGES_REQUESTED** [review]
So firstly - once you finish with the changes I requested, just delete your migration files before making the PR. I'll just regenerate them on the main branch in this instance.

Otherwise, looks solid. A few stylistic things need to be updated (public injections, following naming conventions for those injections, code-behind files) and I had one or two questions about why you designed something the way you did (just curious, honest).

Lemme know if you have any questions about my requested changes.

File: ProjectDataCore/ProjectDataCore.csproj [hunk 11]
```diff
@@ -10,6 +10,10 @@
     <Exec Command="npx postcss $(ProjectDir)obj\$(ConfigurationName)\$(TargetFramework)\scopedcss\projectbundle\$(ProjectName).bundle.scp.css -r" />
     <Exec Command="npx postcss wwwroot/css/app.css -o wwwroot/css/app.min.css" />
   </Target>
+
+  <ItemGroup>
+    <Compile Include="..\ProjectDataCore.Components\Nav\EditModuleDisplay.razor.cs" Link="EditModuleDisplay.razor.cs" />
+  </ItemGroup>
```
Comment (Soyvolon) [inline ProjectDataCore/ProjectDataCore.csproj:16]: Delete this - its causing your error.

File: ProjectDataCore.Components/Nav/EditModuleDisplay.razor.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ProjectDataCore.Components.Nav
+{
+    public partial class EditModuleDisplay
+    {
+        [Parameter] public NavModule Module { get; set; }
+        [Parameter] public Action<NavModule> OnDblClick { get; set; }
+        [Parameter] public Action<NavModule> OnLeftClick { get; set; }
```
Comment (Soyvolon) [inline ProjectDataCore.Components/Nav/EditModuleDisplay.razor.cs:14]: Please make these like the following for consistency:
```
[Parameter]
public type name { get; set; }
```

File: ProjectDataCore.Data/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 13]
```diff
@@ -263,7 +263,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
```
Comment (Soyvolon) [inline ProjectDataCore.Data/Migrations/ApplicationDbContextModelSnapshot.cs:263]: Go ahead and get rid of this file as well.

File: ProjectDataCore.Data/Services/Nav/NavModuleService.cs [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            while (navModules.TryDequeue(out var module))
+            {
+                obj = _dbContext.Entry(module);
+                await obj.Collection(e => e.SubModules)
+                    .LoadAsync();
+
+                foreach (var t in module.SubModules)
+                    navModules.Enqueue(t);
+            }
+        return;
```
Comment (Soyvolon) [inline ProjectDataCore.Data/Services/Nav/NavModuleService.cs:46]: Don't think this return statement is explicitly needed.

File: ProjectDataCore.Data/Services/Nav/NavModuleService.cs [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        _dbContext.Remove(module);
+        _dbContext.SaveChanges();
+        return new(true);
+    }
+
+    public async Task<ActionResult> UpdateNavModuleAsync(NavModule navModule)
+    {
+        if (navModule == null)
+        {
+            return new(false, new() {"paramater is null"});
```
Comment (Soyvolon) [inline ProjectDataCore.Data/Services/Nav/NavModuleService.cs:101]: Capitalize the p in parameter here please - these will get displayed to the user at some point.

File: ProjectDataCore/Shared/NavMenu.razor.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+namespace ProjectDataCore.Shared;
+
+public partial class NavMenu : ComponentBase, IDisposable
+{
+    private List<NavModule> _modules = new();
+
+    public string URI { get; set; } = "";
+
+    [Inject] public NavigationManager NavManager { get; set; }
+    [Inject] public INavModuleService NavModuleService { get; set; }
```
Comment (Soyvolon) [inline ProjectDataCore/Shared/NavMenu.razor.cs:14]: Please change these to the following for consistency:
```
[Inject]
public type name { get; set; }
```

**Review by Soyvolon — CHANGES_REQUESTED** [review]
Just a few minor stylistic changes.

**Review by Soyvolon — APPROVED** [review]
